kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Worker.java

```diff
@@ -27,37 +27,278 @@
+  }
+
+  public static class BaseWriterForTable implements WriterForTable {
+
+    private final IcebergWriterFactory writerFactory;
+    private final Map<String, RecordWriter> writers;
+
+    BaseWriterForTable(IcebergWriterFactory writerFactory, Map<String, RecordWriter> writers) {
+      this.writerFactory = writerFactory;
+      this.writers = writers;
+    }
+
+    @Override
+    public void write(String tableName, SinkRecord record, boolean ignoreMissingTable) {
+      writers
+          .computeIfAbsent(
+              tableName,
+              notUsed -> writerFactory.createWriter(tableName, record, ignoreMissingTable))
+          .write(record);
+    }
+
+    @Override
+    public void writeFailed(
+        SinkRecord sample, String location, Throwable error, String targetTableName) {
+      throw new IllegalArgumentException("BaseWriterForTable cannot write failed records", error);
+    }
+  }
+
+  public static class DeadLetterWriterForTable implements WriterForTable {
+    private static final String PAYLOAD_KEY = "transformed";
+    private static final String ICEBERG_TRANSFORMATION_LOCATION = "ICEBERG_TRANSFORM";
+    private final IcebergWriterFactory writerFactory;
+    private final Map<String, RecordWriter> writers;
+    private final String deadLetterTableName;
+
+    private final String rowIdentifier;
+
+    DeadLetterWriterForTable(
+        IcebergWriterFactory writerFactory,
+        Map<String, RecordWriter> writers,
+        IcebergSinkConfig config) {
+      this.writerFactory = writerFactory;
+      this.writers = writers;
+      Preconditions.checkNotNull(
+          config.deadLetterTableName(), "Dead letter table name cannot be null");
+      Preconditions.checkArgument(
+          !config.deadLetterTableName().isEmpty(), "Dead letter table name cannot be empty");
+      this.deadLetterTableName = config.deadLetterTableName().toLowerCase();
+      this.rowIdentifier = config.connectorName().toLowerCase().replace('-', '_');
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void write(String tableName, SinkRecord record, boolean ignoreMissingTable) {
+      if (record.value() != null) {
+
+        RecordWriter writer = null;
+        SinkRecord recordToWrite = null;
+
+        if (record.value() instanceof Map) {
+          Map<String, Object> payload = (Map<String, Object>) record.value();
+          SinkRecord transformed = (SinkRecord) payload.get(PAYLOAD_KEY);
+          try {
+            writer =
+                writers.computeIfAbsent(
+                    tableName,
+                    notUsed ->
+                        writerFactory.createWriter(tableName, transformed, ignoreMissingTable));
+            recordToWrite = transformed;
+          } catch (DeadLetterUtils.DeadLetterException error) {
+            writeFailed(record, error.getLocation(), error.getError(), tableName);
+          }
```

Comment:
You could.  If the record is a map (which it can be for a lot of payloads) it's by definition not a failed record.

A failed record is always a struct which is why that check comes later.

Reply:
Got it, makes sense.